src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/CompilerScope.cs

```diff
@@ -462,19 +462,23 @@ private void AllocateLocals(LambdaCompiler lc)
             }
         }
 
-        private IList<ParameterExpression> GetVariables()
+        private IEnumerable<ParameterExpression> GetVariables() =>
+            MergedScopes == null ? GetVariables(Node) : GetVariablesIncludingMerged();
+
+        private IEnumerable<ParameterExpression> GetVariablesIncludingMerged()
         {
-            var vars = GetVariables(Node);
-            if (MergedScopes == null)
+            foreach (ParameterExpression param in GetVariables(Node))
             {
-                return vars;
+                yield return param;
             }
-            var list = new List<ParameterExpression>(vars);
+
             foreach (var scope in MergedScopes)
             {
-                list.AddRange(GetVariables(scope));
+                foreach (ParameterExpression param in scope.Variables)
```

Comment:
individual scopes tend to be small. often just one variable. the change makes sense